### ISSUE TITLE
Loft should deal with finance company payments made before liability transfer

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2250]
_Imported from trac ticket 2250.
Created by **kbower**
Opened in trac: Mar 25, 2013, 7:59:25 AM
Last modified in trac: Mar 27, 2013, 5:16:06 PM_
***
This happens at Ivan Smith.

Unfortunately, this probably has extended implications with:
 * client
 * server money movement framework

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 25, 2013, 8:01:57 AM:_

(Changeset [[3955]](../commit/7416ef74dc7c9f410f7f94cdfc0bfa9e4a049940))
* See #2250, make Ivan Smith liability transfer always be tentativefinancedamount (this won't be accurate always if order split after some payments are made)

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 25, 2013, 8:04:51 AM:_

The whole 448.07 should be transferred to the finance company as liability.

We've had a communication problem here, which has shot down a bunch of the programming we wrote for Loft.  A few months back I asked all our clients whether liability transfer should ever occur before the order is finalized.  All, including Ivan Smith, said "no."  My (apparently incorrect) assumption was that there is no way we will start getting payments for the order before the liability was transferred.  It seemed (and still seems) to me that should be impossible.  If we model the real world, then, once the financier assumes liability, we record it in the system (we don't wait until the order is closed).  Also, clearly, so I thought, we can't have any payments coming from the finance company before they've even assumed liability.

Here, I see you are getting payments before the liability is transferred.  However, currently Loft is programmed to only transfer liability of the lesser amount of the outstanding balance and the tentativefinancedamount.  This is to safe guard against accidentally financing an amount that includes items that were voided since the tentativefinancedamount was recorded.

Here, it is backfiring, because essentially Loft is saying "well, it's time to transfer liability; at most transfer the orderbalance", which orderbalance Loft assumes can't possibly reflect early payments from the financier.

So, ideally, we need to make Loft factor out "early" payments when it transfers liability.  (Really, we need to rewrite other aspects now, given the knowledge that this happens.)

Alternatively, as a quick fix, we can add this rule to Ivan Smith's start up rule:

```
def isf_amounttofinance(order):
    return max(Decimal(0), order.tentativefinancedamount)
Order._amounttofinance = isf_amounttofinance
```
This will mean we always request funding for and transfer liability for the full tentativefinancedamount.  It seems less desirable than factoring out "early" payments, though, because we lose the safety net that guards against financing voided lines.  If Ivan Smith doesn't care for that safety net, then that rule can be the permanent fix (though I still think we should fix Loft with this new understanding).

Another side effect you will notice is that as soon as the first payment comes in, the "financedamount" will appear to drop from the tentativefinancedamount down to the first payment amount.  This is because Loft doesn't know that we expect to transfer liability, so it must assume that that single payment is all we are getting from the financier (caters to other retailers who don't transfer liability).  In retrospect, with a better understanding now, we should be able to rely on the system parameter to tell Loft "no, they intend to transfer liability, so continue to display the tentativefinancedamount" as the financedamount.

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 25, 2013, 8:11:32 AM:_

isf: 0313342QE62

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 27, 2013, 5:16:06 PM:_

(Changeset [[3980]](../commit/58781c4e7526318d560729f8664c70445f139ad2))
* fixes #2250, Loft should deal with finance company payments made before liability transfer #isf

**Comment by kentbower** [comment]
_**kbower** trac status change on Mar 27, 2013, 5:16:06 PM:_ **closed**
